app/controllers/users_controller.rb

```diff
@@ -83,10 +84,10 @@ def persisted_form_attributes
   end
 
   def group_ids
-    current_user.groups.map { |group| { group_id: group.id } }
+    current_user.groups.map { |group| group.id }
   end
 
   def role_names
-    current_user.roles.map { |role| { name: role.name } }
+    current_user.roles.map { |role| role.name }
```

Comment:
こちらも
https://github.com/IN00/self_introduction/pull/142/files#r430324526
同様に
```suggestion
    current_user.roles.map(&:name)
```
もしくは
```ruby
current_user.roles.pluck(:name)
```

Reply:
というか、group_idsの順番とrole_namesの配列の順番は必ず一致することが保証されないよ？大丈夫？

Reply:
> SQLではORDER BYしない限り順番は保証されない

ちょっと調べてみます。

Reply:
> UserRoleGroupの中間テーブルを介して, roles, groupsをそれぞれcurrent_userから呼んでいるので順番は保証されると思っていました

2つの理由で違います。

1. MySQLの仕様上

MySQLでORDER BYをつけないときの並び順 https://blog.kamipo.net/entry/2016/12/24/234944
> 並び順が実装と実行計画に依存する

なので**不定**です。
今回InnoDBを使っているとしたら、主キー順でくることもあれば、他にINDEXが貼られていてそれが使われたらそちらの順で来ることもという感じでMySQLの心次第です。

2. ActiveRecordの仕様上

ちゃんと実行されるSQL見てください。
たとえば
```ruby
class Prefecture < ApplicationRecord
  has_many :cities
  has_many :areas, through: :cities
end

class City < ApplicationRecord
  belongs_to :prefecture
end

class Area < ApplicationRecord
  belongs_to :city
end
```
とあったとして、
```ruby
pry(main)> prefecture = Prefecture.find(13)
pry(main)> prefecture.cities.map(&:id)
  City Load (1.1ms)  SELECT `cities`.* FROM `cities` WHERE `cities`.`prefecture_id` = 13
pry(main)> prefecture.areas.map(&:id)
  Area Load (3.1ms)  SELECT `areas`.* FROM `areas` INNER JOIN `cities` ON `areas`.`city_id` = `cities`.`id` WHERE `cities`.`prefecture_id` = 13
```
とthroughしてても中間テーブルはJOIN句で使われるのみで都度SQLが実行されます
MySQLの仕様上順番は不定です。

Reply:
ちなみに2回目移行はCityのActiveRecordのオブジェクトが既に作られているので
```ruby
pry(main)> prefecture.cities.map(&:id)
```
と実行しても再度SQLは実行されません（ `cities` にArrayとして既に入ってるのをmapで取り出すイメージ ）
一方 `.ids` は都度SQLを実行します。 `pluck(:id)` と同じです。ActiveRecordのオブジェクトは生成されません（なのでその分1回あたりの実行時間は速い）

Reply:
今回の仕様では、おそらく最初の入力フォームで、
`{ group_id: 4, role_name: 'any' # role.id: 1 }`
次の入力フォームで
`{ group_id: 1, role_name: 'any' # role.id: 2 }`

となったら、特に関連するINDEXなさそうだから
```ruby
current_user.groups.map(&:id) # => [1, 4]
current_user.roles.map(&:id) # => [1, 2]
```
主キーの昇順でそれぞれ返ってきそうだけどどうかな？

Reply:
あ、HerokuでPostgreSQLにするのか

PostgreSQL 12.0文書 https://www.postgresql.jp/document/12/html/sql-select.html
> ORDER BY句が指定された場合、返される行は指定した順番でソートされます。 ORDER BYが指定されない場合は、システムが計算過程で見つけた順番で行が返されます （後述のORDER BY句を参照してください）。

こちらもMySQLと同様実装依存っぽい

Reply:
なるほどindexが貼られていれば順番が違ってくる可能性があるんですね。

> とthroughしてても中間テーブルはJOIN句で使われるのみで都度SQLが実行されます
MySQLの仕様上順番は不定です。

なるほど。ORDER BYしない限り順番が保証されないならば、ActiveRecordの仕様上、中間テーブルを介するだけでは発行されるSQLをみた感じ保証されてないですね。

> となったら、特に関連するINDEXなさそうだから
>
>```
>current_user.groups.map(&:id) # => [1, 4]
>current_user.roles.map(&:id) # => [1, 2]
>```
>主キーの昇順でそれぞれ返ってきそうだけどどうかな？

それでいうと、
```
[2] pry(main)> user.groups.map(&:id)
=> [7, 3]                               
[3] pry(main)> user.roles.map(&:id)
=> [7, 8]
```

主キーの昇順ではなく、一応順番は保持されてはいました。ただこの順番がいつも保持される保証がないという事ですよね？
いつこの順番が崩れるかような変更が加わるかわからないので今のうちに順番ほ保持する修正を加えます。

Reply:
`current_user.groups.map(&:id)`
と
`current_user.roles.map(&:id)`
それぞれどういうSQLが実行されてますか？

Reply:
あとレコードの内容がそれぞれ知りたい

Reply:
```ruby
group_ids = current_user.user_role_groups.map { |urg| urg.group_id }
role_ids = current_user.user_role_groups.map { |urg| urg.role_id }
```
が良いのでは説
外部キー制約張ってあってNOT NULLでバリデーションもやってるならこれでよさげなよーな

Reply:
```java
[2] pry(main)> user.groups.map(&:id)
  Group Load (0.3ms)  SELECT `groups`.* FROM `groups` INNER JOIN `user_role_groups` ON `groups`.`id` = `user_role_groups`.`group_id` WHERE `user_role_groups`.`user_id` = 24
=> [7, 3]
[3] pry(main)> user.roles.map(&:id)
  Role Load (0.3ms)  SELECT `roles`.* FROM `roles` INNER JOIN `user_role_groups` ON `roles`.`id` = `user_role_groups`.`role_id` WHERE `user_role_groups`.`user_id` = 24
=> [7, 8]
```

Reply:
> あとレコードの内容がそれぞれ知りたい

```java
[4] pry(main)> Group.all
  Group Load (3.8ms)  SELECT `groups`.* FROM `groups`
=> [
 #<Group:0x00007fde71f8bd10 id: 1, name: "Space owner", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8bbf8 id: 2, name: "Online sales", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8bae0 id: 3, name: "POP-UP NOW", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b770 id: 4, name: "Marketing", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b680 id: 5, name: "Coporate", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b5b8 id: 6, name: "adpt OS", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b4f0 id: 7, name: "product", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b400 id: 8, name: "CEO", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b298 id: 9, name: "CFO", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Group:0x00007fde71f8b130 id: 10, name: "CDO", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>]
```

```java
[6] pry(main)> Role.all
  Role Load (0.8ms)  SELECT `roles`.* FROM `roles`
=> [
 #<Role:0x00007fdea19ba140 id: 1, name: "社長です。気軽に話しかけてください。",                  created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Role:0x00007fdea19ba050 id: 2, name: "俺か、俺以外か",                                    created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Role:0x00007fdea19b9ee8 id: 3, name: "今までにないサービスを作り、POPUPを盛り上げていきます！", created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Role:0x00007fdea19b9cb8 id: 4, name: "スペースの価値を最大化させるために頑張っていいます!",     created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Role:0x00007fdea19b9a38 id: 5, name: "○○を実現する事をしています。",                        created_at: Thu, 28 May 2020 00:53:25 UTC +00:00, updated_at: Thu, 28 May 2020 00:53:25 UTC +00:00>,
 #<Role:0x00007fdea19b98d0 id: 7, name: "○○を実現する事をしています。",                        created_at: Thu, 28 May 2020 01:32:48 UTC +00:00, updated_at: Thu, 28 May 2020 01:32:48 UTC +00:00>,
 #<Role:0x00007fdea19b9678 id: 8, name: "pop up now",                                      created_at: Thu, 28 May 2020 01:32:48 UTC +00:00, updated_at: Thu, 28 May 2020 01:32:48 UTC +00:00>]
```

Reply:
> 主キーの昇順ではなく、一応順番は保持されてはいました。

ちゃんと伝わってなかったら申し訳ないんですが、僕が言っているのは、並び順が実装と実行計画に依存するのであくまでも今の実装上 **一応** 、**たまたま**この順番になっているだけで、他で修正入ったら変わる可能性あるよってことです。ほんと一応ですね。

今は現時点での実装上効いているINDEXやら主キーやらで主キーの順番で返ってきてます。そうすると他で修正、たとえばroleを毎回作らず使い回すことになって、user_role_groupsのrole_idが主キーの順番通りに入ってこなくなり、
```ruby
[
 #<UserRoleGroup:0x00007fdea1b000e0 id: 31, user_id: 24, role_id: 8, group_id: 7, created_at: Thu, 28 May 2020 01:32:48 UTC +00:00, updated_at: Thu, 28 May 2020 01:32:48 UTC +00:00>,
 #<UserRoleGroup:0x00007fdea1ada2f0 id: 32, user_id: 24, role_id: 7, group_id: 3, created_at: Thu, 28 May 2020 01:32:48 UTC +00:00, updated_at: Thu, 28 May 2020 01:32:48 UTC +00:00>]
```
みたいなデータとなった場合、途端にgroupsとrolesの順番がずれるかと。

適当に作ったデータだけど、たとえばrole_idが主キーの順番通りに入らなかった場合、

```sql
mysql> SELECT * FROM `groups` INNER JOIN `user_role_groups` ON `groups`.`id` = `user_role_groups`.`group_id` WHERE `user_role_groups`.`user_id` = 18354;
+----+------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+
| id | name | created_at                 | updated_at                 | id | user_id    | role_id   | group_id   | created_at                 | updated_at                 |
+----+------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+
|  6 | あ   | 2020-05-28 06:28:15.551860 | 2020-05-28 06:28:15.551860 |  1 |      18354 |         2 |          6 | 2020-05-28 06:29:45.796013 | 2020-05-28 06:29:45.796013 |
|  5 | DDD  | 2020-05-28 06:28:13.213337 | 2020-05-28 06:28:13.213337 |  2 |      18354 |         1 |          5 | 2020-05-28 06:29:50.890897 | 2020-05-28 06:29:50.890897 |
|  1 | bbb  | 2020-05-28 06:27:53.520047 | 2020-05-28 06:27:53.520047 |  3 |      18354 |         3 |          1 | 2020-05-28 06:29:56.670504 | 2020-05-28 06:29:56.670504 |
|  2 | ccc  | 2020-05-28 06:27:56.944151 | 2020-05-28 06:27:56.944151 |  6 |      18354 |         4 |          2 | 2020-05-28 06:31:01.032211 | 2020-05-28 06:31:01.032211 |
+----+------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+
```
```sql
# current_user.roles.map(&:id)
mysql> SELECT * FROM `roles` INNER JOIN `user_role_groups` ON `roles`.`id` = `user_role_groups`.`role_id` WHERE `user_role_groups`.`user_id` = 18354;
+----+-----------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+
| id | name      | created_at                 | updated_at                 | id | user_id    | role_id   | group_id   | created_at                 | updated_at                 |
+----+-----------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+
|  1 | bbbbbrole | 2020-05-28 06:28:55.501970 | 2020-05-28 06:28:55.501970 |  2 |      18354 |         1 |          5 | 2020-05-28 06:29:50.890897 | 2020-05-28 06:29:50.890897 |
|  2 | cccccrole | 2020-05-28 06:29:01.384182 | 2020-05-28 06:29:01.384182 |  1 |      18354 |         2 |          6 | 2020-05-28 06:29:45.796013 | 2020-05-28 06:29:45.796013 |
|  3 | 1         | 2020-05-28 06:29:05.546496 | 2020-05-28 06:29:05.546496 |  3 |      18354 |         3 |          1 | 2020-05-28 06:29:56.670504 | 2020-05-28 06:29:56.670504 |
|  4 | 444444    | 2020-05-28 06:30:56.060342 | 2020-05-28 06:30:56.060342 |  6 |      18354 |         4 |          2 | 2020-05-28 06:31:01.032211 | 2020-05-28 06:31:01.032211 |
+----+-----------+----------------------------+----------------------------+----+------------+-----------+------------+----------------------------+----------------------------+

# current_user.roles.ids
mysql> SELECT `roles`.`id` FROM `roles` INNER JOIN `user_role_groups` ON `roles`.`id` = `user_role_groups`.`role_id` WHERE `user_role_groups`.`user_id` = 18354;
+----+
| id |
+----+
|  2 |
|  1 |
|  3 |
|  4 |
+----+
```
と少なくとも僕のローカルのMySQLの環境では `map` と `ids` で挙動違ったりするようになります。
（EXPLAIN見る限りインデックスの効き方が変わってて、idsはidだけ取るからcovering indexでインデックスだけで完結してるから順番変わってるのだと思う。）
https://techlife.cookpad.com/entry/2017/04/18/092524

これは修正者にとって気づきづらいバグです。
まだORDER BYしてて、ここORDER BYしてるからバグってるのか！って気づけた方がマシで、これ実際遭遇したら調査が大変そう。
順番が保たれているかのテスト書いてたら気づけるけど、時限爆弾みたいなもんなんで、今から想定できるものは潰しておきたいです。
運に身を任せて変更に弱いナイーブな実装はしてほしくないですね。

Reply:
再度だけど、groupとroleの順序が保証されてる `user_role_groups` から取ればいいんじゃないかな？

Reply:
> 少なくとも僕のローカルのMySQLの環境では map と ids で挙動違ったりするようになります。

なるほど、志村さんでの環境で挙動違うことから納得しました。

> group_ids = current_user.user_role_groups.map { |urg| urg.group_id }
role_ids = current_user.user_role_groups.map { |urg| urg.role_id }
が良いのでは説
外部キー制約張ってあってNOT NULLでバリデーションもやってるならこれでよさげなよーな

> 再度だけど、groupとroleの順序が保証されてる user_role_groups から取ればいいんじゃないかな？

↓で修正しました！

Reply:
> なるほど、志村さんでの環境で挙動違うことから納得しました。

この文言だけ見るとちゃんと理解いただけてないように見受けられるのだけど、環境もそうだけど、SQL文とかレコード数とかオプティマイザー次第です。
この指摘の本質は、順序が保証される必要がある要件なので、実装上冪等性を持っていないといけない点です。
まずなんとなくでいいから `オプティマイザー` とか `実行計画` って分かる？単語の意味調べた？

Reply:
修正自体は良いと思います。あとは理解をしてほしいです。